### PR TITLE
[WIP][Flax] CLIP training example

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/flax/vision/run_clip_flax.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+has_tensorboard = is_tensorboard_available()
+if has_tensorboard:
+    try:
+        from flax.metrics.tensorboard import SummaryWriter
+    except ImportError as ie:
+        has_tensorboard = False
+        print(f"Unable to display metrics through TensorBoard because some package are not installed: {ie}")
+
+else:
+    print(
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_clip_flax.py:79]: put into main and use log() instead?

File: examples/flax/vision/run_clip_flax.py [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    #  Get the datasets: you can either provide your own CSV/JSON/TXT training and evaluation files (see below)
+    # or just provide the name of one of the public datasets available on the hub at https://huggingface.co/datasets/
+    # (the dataset will be downloaded automatically from the datasets Hub).
+    #
+    # For CSV/JSON files, this script will use the column called 'text' or the first column if no column called
+    # 'text' is found. You can easily tweak this behavior (see below).
+    #
+    # In distributed training, the load_dataset function guarantees that only one local process can concurrently
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_clip_flax.py:339]: comment can be deleted

File: examples/flax/vision/run_clip_flax.py [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+        training_args.learning_rate,
+    )
+
+    # We use Optax's "masking" functionality to not apply weight decay
+    # to bias and LayerNorm scale parameters. decay_mask_fn returns a
+    # mask boolean with the same structure as the parameters.
+    # The mask is True for parameters that should be decayed.
+    def decay_mask_fn(params):
+        flat_params = traverse_util.flatten_dict(params)
+        flat_mask = {path: (path[-1] != "bias" and path[-2:] != ("LayerNorm", "scale")) for path in flat_params}
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_clip_flax.py:478]: should be adapted for the correct weight names

File: examples/flax/vision/run_clip_flax.py [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+    state = TrainState.create(apply_fn=model.__call__, params=model.params, tx=adamw, dropout_rng=dropout_rng)
+
+    def cross_entropy(logits, axis):
+        logprobs = jax.nn.log_softmax(logits, axis=axis)
+        nll = jnp.diag(logprobs)
+        ce = -jnp.mean(nll)
+        return ce
+
+    def clip_loss(similarity):
+        loss = (cross_entropy(similarity, axis=0) + cross_entropy(similarity, axis=1)) / 2
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_clip_flax.py:501]: is it correct that `logits == similarity`?

File: examples/flax/vision/run_clip_flax.py [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+        eps=training_args.adam_epsilon,
+        weight_decay=training_args.weight_decay,
+        mask=decay_mask_fn,
+    )
+
+    # Setup train state
+    state = TrainState.create(apply_fn=model.__call__, params=model.params, tx=adamw, dropout_rng=dropout_rng)
+
+    def cross_entropy(logits, axis):
+        logprobs = jax.nn.log_softmax(logits, axis=axis)
```
Comment (patrickvonplaten) [inline examples/flax/vision/run_clip_flax.py:495]: maybe worth verifying this and integration testing with the original code if some error persists

**Comment by patil-suraj** [comment]
closing this PR, messed up the commit history :( 
Opened a new PR here #12491
